File: src/main/java/org/example/commons/AbstractRepository.java
package org.example.commons;

import java.util.Arrays;
import java.util.List;
import javax.inject.Inject;
import javax.persistence.EntityManager;
import javax.persistence.criteria.CriteriaBuilder;
import javax.persistence.criteria.CriteriaQuery;
import javax.persistence.criteria.Root;
import javax.persistence.metamodel.SingularAttribute;

import org.example.commons.api.Filter;
import org.example.commons.api.Repository;
import org.hibernate.Session;

abstract public class AbstractRepository<T, S> implements Repository<T, S> {

    @Inject
    private EntityManager em;

    private Class<T> cls;

    protected AbstractRepository(Class<T> clazz) {
        cls = clazz;
    }

    protected Session getSession() {
        return em.unwrap(Session.class);
    }

    @Override
    public List<T> findAll() {
        CriteriaBuilder builder = getSession().getCriteriaBuilder();
        CriteriaQuery<T> query = builder.createQuery(cls);
        Root<T> root = query.from(cls);
        query.select(root);
        return getSession().createQuery(query).getResultList();
    }

    @Override
    public T findById(S id) {
        return getSession().find(cls, id);
    }

    @Override
    public List<T> findByIds(S... ids) {
        return findByIds(Arrays.asList(ids));
    }

    @Override
    public List<T> findByIds(List<S> ids) {
        CriteriaBuilder builder = getSession().getCriteriaBuilder();
        CriteriaQuery<T> query = builder.createQuery(cls);
        Root<T> root = query.from(cls);
        query.where(root.get("id").in(ids));
        return getSession().createQuery(query).getResultList();
    }

    @Override
    public <X> List<T> findByColumn(SingularAttribute<T, X> column, X... values) {
        return findByColumn(column, Arrays.asList(values));
    }

    @Override
    public <X> List<T> findByColumn(SingularAttribute<T, X> column, List<X> values) {
        CriteriaBuilder builder = getSession().getCriteriaBuilder();
        CriteriaQuery<T> query = builder.createQuery(cls);
        Root<T> root = query.from(cls);
        query.where(root.get(column).in(values));
        return getSession().createQuery(query).getResultList();
    }

    @Override
    public void create(T... entities) {
        create(Arrays.asList(entities));
    }

    @Override
    public void create(List<T> entities) {
        em.getTransaction().begin();
        entities.forEach(e -> em.persist(e));
        em.getTransaction().commit();
    }

    @Override
    public void update(T... entities) {
        update(Arrays.asList(entities));
    }

    @Override
    public void update(List<T> entities) {
        em.getTransaction().begin();
        entities.forEach(e -> em.merge(e));
        em.getTransaction().commit();
    }

    @Override
    public void delete(T... entities) {
        delete(Arrays.asList(entities));
    }

    @Override
    public void delete(List<T> entities) {
        getSession().getTransaction().begin();
        entities.forEach(e -> getSession().delete(e));
        getSession().getTransaction().commit();
    }

    public abstract List<T> filter(Filter<T> filter);
}

File: src/main/kotlin/org/example/commons/AbstractRepository.kt
package org.example.commons

import org.example.commons.api.Repository
import org.hibernate.Session
import java.util.*
import java.util.function.Consumer
import javax.inject.Inject
import javax.persistence.EntityManager
import javax.persistence.metamodel.SingularAttribute

abstract class AbstractRepository<T, S> protected constructor(private val cls: Class<T>) : Repository<T, S> {
    @Inject
    open lateinit var em: EntityManager

    protected val session: Session
        get() = em.unwrap(Session::class.java)

    override fun findAll(): MutableList<T> {
        val builder = session.criteriaBuilder
        val query = builder.createQuery(cls)
        val root = query.from(cls)
        query.select(root)
        return session.createQuery(query).resultList
    }

    override fun findById(id: S): T {
        return session.find(cls, id)
    }

    override fun findByIds(vararg ids: S): MutableList<T> {
        return findByIds(listOf(*ids))
    }

    override fun findByIds(ids: Iterable<S>): MutableList<T> {
        val builder = session.criteriaBuilder
        val query = builder.createQuery(cls)
        val root = query.from(cls)
        query.where(root.get<Any>("id").`in`(ids))
        return session.createQuery(query).resultList
    }

    override fun <X> findByColumn(column: SingularAttribute<T, X>, vararg values: X): MutableList<T> {
        return findByColumn(column, listOf(*values))
    }

    override fun <X> findByColumn(column: SingularAttribute<T, X>, values: Iterable<X>): MutableList<T> {
        val builder = session.criteriaBuilder
        val query = builder.createQuery(cls)
        val root = query.from(cls)
        query.where(root[column].`in`(values))
        return session.createQuery(query).resultList
    }

    override fun create(vararg entities: T) {
        create(listOf(*entities))
    }

    override fun create(entities: Iterable<T>) {
        em.transaction.begin()
        entities.forEach { e: T -> em.persist(e) }
        em.transaction.commit()
    }

    override fun update(vararg entities: T) {
        update(listOf(*entities))
    }

    override fun update(entities: Iterable<T>) {
        em.transaction.begin()
        entities.forEach { e -> em.merge(e) }
        em.transaction.commit()
    }

    override fun delete(vararg entities: T) {
        delete(listOf(*entities))
    }

    override fun delete(entities: Iterable<T>) {
        em.transaction.begin()
        entities.forEach { e -> em.remove(e) }
        em.transaction.commit()
    }

}